spec/controllers/checkout_controller_concurrency_spec.rb

```diff
@@ -42,12 +42,15 @@
     allow(controller).to receive(:spree_current_user).and_return(order.user)
     allow(controller).to receive(:current_distributor).and_return(order.distributor)
     allow(controller).to receive(:current_order_cycle).and_return(order.order_cycle)
+
+    # Avoid setting headers as Rails 5 is not thread-safe here:
+    allow(controller).to receive(:restrict_iframes)
   end
 
   # This spec does not seem to be running in two threads in Rails 5. There are errors for the
   # same response headers being set twice, possibly indicating that there is only one response
   # as opposed to two separate requests in two threads?
-  xit "handles two concurrent orders successfully" do
+  it "handles two concurrent orders successfully" do
```

Comment:
we need to work on making this test more resilient if we want to enable it back. It's very very flaky.

Reply:
I think I disabled this during the Rails 5 upgrade because it was failing ~90% of the time...

Reply:
I think I commented this out. I spent some time on this test and my conclusion after some serious debugging is that this test was invalid, the breakpoint was not working so the threads were just running one after the other all the time.

Reply:
Thank you all for raising your points. That made me double-check:

> It's very very flaky.

First, I tested if the the spec fails reliably. I commented out the pessimistic locking and ran the spec a dozen times. It always failed. :heavy_check_mark: 

Second, I ran the code that should pass 100 times. It passed 100 times. :heavy_check_mark: 

Third, I ran the specs 20 more times with my CPU maxed out with other processes. They all passed again. :heavy_check_mark: 

> my conclusion after some serious debugging is that this test was invalid, the breakpoint was not working so the threads were just running one after the other all the time.

I had to stub `#restrict_iframes` because the app ended up in an invalid state. It tried to add headers in the second thread when the first sent them already. That made me think that the requests are executed concurrently.

I read through the code and noticed that the mutex lock wasn't used though. I'm surprised that the spec failed so reliably without the lock. It's lucky but it could accidentally pass even if the code is not syncing properly.

Looking through the history I found commit b0fa1464f of PR #6119 which took the usage of the breakpoint out. @sauloperez and @luisramos0 approved that PR without realising that it would make the spec unreliable. And that's totally understandable because that code was using a random, unrelated but convenient method to synchronise the threads. I wonder how we can make the test easier to understand.

The purpose of the lock is to ensure that both threads load the order and complete all the before_action checks first. And only after both threads did all the checks and are waiting for the lock, we release the lock and they both try to process the checkout at the same time with stale objects. That is the flow when the test is failing.

If the controller uses pessimistic locking correctly, then the second thread is waiting before any of the before_action checks. Those checks run after the first thread finished and see the processed order correctly.

I re-added the synchronisation via the mutex lock to make the spec reliably fail (even though it was doing that on my machine already). I then re-ran 100 tests passing and 100 test failing. It worked as expected. The intentional failures were quite varied because the two threads race unpredictably. Here are some examples:

```
     StateMachines::InvalidTransition:
       Cannot transition state via :restart_checkout from :complete (Reason(s): State cannot transition via "restart checkout")

     AbstractController::DoubleRenderError:
       Render and/or redirect were called multiple times in this action. Please note that you may only call render OR redirect, and at most once per action. Also note that neither redirect nor render terminate execution of the action, so if you want to exit an action after redirecting, you need to do something like "redirect_to(...) and return".

     ActiveRecord::StatementInvalid:
       PG::TRDeadlockDetected: ERROR:  deadlock detected
```

The DoubleRenderError is probably a bug in our code but out of scope here. Please have a look at the new commit and let me know if you understand it and if we can make it easier to understand. I was able to use a newer rspec syntax now that's hopefully clearer.

Reply:
ups, yes, we totally missed that PR removing the syncing :see_no_evil: